### PR TITLE
382 raise error for duplicate plugins

### LINKED ISSUE #230: Duplicate plugins with the same ID should raise a PluginError
# Requested Update

### Description
Plugin IDs should be required to be unique within its interface across all packages. When a non-unique plugin ID is encountered a PluginError should be raised.

The error should state:
- interface name
- plugin ID
- the names of all packages where that plugin ID was encountered
  - I expect that this will error as soon as it encounters a conflict, so probably just one or two packages

Note that I'm saying "ID" rather than "name". This is because some interfaces may have plugins whose names conflict but whose IDs are more complex. An example of this is the Products interface whose plugin IDs are tuples containing `name` and `source_name`.

### Checklist for Completion
* [ ]

### LINKED ISSUE #238: Generate Plugin list during installation
# Requested Update

### Description
This issue has been created to generate all available geoips plugins during installation. The purpose of this issue is to improve the loading time during run_procflow, which currently loads all available plugins each time such process is ran. Another issue related to setuptools_scm will work on calling this script

### LINKED ISSUE #382: Add Support for Raising an error when two Plugins have the same name
# Requested Update

### Description
It's ok to name plugins the same, provided they are in different interfaces. However, when two plugins have the same name in the same interface, GeoIPS will overwrite the first accessed plugin with the new Plugin attributes. To avoid this, we should ensure that plugins with the same name in the same interfaces raise an error when encountered.

### Background and Motivation
This was discovered when making the "create_plugin_registry" branch (see #240). The function "merge_nested_dictionary" in geoips_utils.py will overwrite the aforementioned plugins if that case occurs. We should raise an error to fix this.

### Code to demonstrate issue
See this [function](https://github.com/NRLMMD-GEOIPS/geoips/blob/0dd8dfbca1fb915a280d747e0edef000dc31c15a/geoips/geoips_utils.py#L499C16-L499C16) for the code that hosts this issue.

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] Required ***unit tests*** added and pass for new/modified functionality
* [x] NO REQUIRED ***integration tests*** (explain why not required)
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [x] NO REQUIRED ***updates to other repos*** (explain why not required)

# Related Issues
fixes #382
fixes #238 
fixes #230 (I think)
branched off of #240 

# Testing Instructions
To test this, change the name of a plugin within the same interface as another plugin to the same thing. The functionality added will catch this error and not allow you to move on until you've fixed those duplicates.

# Summary
Currently in GeoIPS, if two plugins of the same interface have the same name, one is
just overwritten. This would occur in the order in which GeoIPS defines its entry
points. This is a bug, and needs to be fixed. To do so, we check all
registered_plugins.yaml found in each package, loop through them, and compare them
against every other plugin registry. If duplicates are found, an error is raised, and
the registries are removed until one of the duplicates is renamed.

    modified: geoips/geoips/create_plugin_registries.py

# Output
If two plugins of the same interface, of the same name, are found in the same package, this type of error is raised:
```
ERROR: Removing registries due to duplicate plugins. You must fix the error shown below before GeoIPS can operate correctly.
Once fixed, please run 'create_plugin_registries' to set up GeoIPS appropriately



Traceback (most recent call last):
  File "/home/evrose54/anaconda3/envs/geoips/bin/create_plugin_registries", line 8, in <module>
    sys.exit(main())
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 438, in main
    create_plugin_registries(plugin_packages)
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 236, in create_plugin_registries
    parse_plugin_paths(plugin_paths, package, pkg_dir, plugins)
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 273, in parse_plugin_paths
    add_yaml_plugin(filepath, relpath, package, plugins["yaml_based"])
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 306, in add_yaml_plugin
    check_plugin_exists(package, plugins, interface_name, plugin["name"])
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 164, in check_plugin_exists
    raise PluginRegistryError(
geoips.errors.PluginRegistryError: Error in package [geoips]:
            You can not have two Plugins of the same interface [sectors]
            with the same name [australia].
```

If two plugins of the Products Interface, of the same name, are found in two different packages, this type of error is raised:

```
ERROR: Removing registries due to duplicate plugins. You must fix the error shown below before GeoIPS can operate correctly.
Once fixed, please run 'create_plugin_registries' to set up GeoIPS appropriately



Traceback (most recent call last):
  File "/home/evrose54/anaconda3/envs/geoips/bin/create_plugin_registries", line 8, in <module>
    sys.exit(main())
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 438, in main
    create_plugin_registries(plugin_packages)
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 246, in create_plugin_registries
    registry_sanity_check(plugin_packages)
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 134, in registry_sanity_check
    raise PluginRegistryError(
geoips.errors.PluginRegistryError: Error with packages [geoips, my_package]:
                                            You can't have two products of the same
                                            interface [products] with the same name [89pct] found
                                            under source name [amsr2].
```

If two plugins of the same interface, of the same name, in a package outside of geoips are found, this type of error is raised.

```
ERROR: Removing registries due to duplicate plugins. You must fix the error shown below before GeoIPS can operate correctly.
Once fixed, please run 'create_plugin_registries' to set up GeoIPS appropriately



Traceback (most recent call last):
  File "/home/evrose54/anaconda3/envs/geoips/bin/create_plugin_registries", line 8, in <module>
    sys.exit(main())
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 441, in main
    create_plugin_registries(plugin_packages)
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 239, in create_plugin_registries
    parse_plugin_paths(plugin_paths, package, pkg_dir, plugins)
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 282, in parse_plugin_paths
    add_module_plugin(package, relpath, plugins["module_based"])
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 424, in add_module_plugin
    check_plugin_exists(package, plugins, interface_name, name)
  File "/local/home/evrose54/geoips/geoips_packages/geoips/geoips/create_plugin_registries.py", line 167, in check_plugin_exists
    raise PluginRegistryError(
geoips.errors.PluginRegistryError: Error in package [recenter_tc]:
            You can not have two Plugins of the same interface [filename_formatters]
            with the same name [archer_fix].
```

### CODE REVIEW

File: docs/source/releases/v1_11_7a0.rst
Comment (mindyls) [inline docs/source/releases/v1_11_7a0.rst:26]: NOTE: do not merge this until #238 has been merged.  When #238 has been merged to main, this one will automatically update so the destination is "main"

File: geoips/create_plugin_registries.py
Comment (mindyls) [inline geoips/create_plugin_registries.py:62]: Could you comment in the code a bit what all these loops are doing?  ie, comp_idx, pkg_idx, etc?

File: geoips/create_plugin_registries.py
Comment (mindyls) [inline geoips/create_plugin_registries.py:39]: So this will remove all registries during install?  We will just need to make sure the install fails with a useful error message when we get to the point of tying this in with pyproject.toml?

File: geoips/create_plugin_registries.py
Comment (mindyls) [inline geoips/create_plugin_registries.py:107]: Could you include copy/paste of some of the output when you have duplicate plugin names in the body of the PR please?  Ie, with the same name in 2 separate repos, same name both in the geoips repo, same name both in another repo maybe?

**Review by mindyls — COMMENTED** [review]
Please let me know when my conversations have been addressed and let me know when I should re-review!

File: geoips/create_plugin_registries.py
Comment (evrose54) [reply geoips/create_plugin_registries.py:62]: Yup! I can add some more comments.

File: geoips/create_plugin_registries.py
Comment (evrose54) [reply geoips/create_plugin_registries.py:39]: Yup! If a duplicate is found all registries are removed. I've added a few error messages that should point the user to fix those duplicates, but if you'd like me to modify them just let me know. These errors also state that you can't use GeoIPS functionality until these errors are fixed.

File: geoips/create_plugin_registries.py
Comment (evrose54) [reply geoips/create_plugin_registries.py:107]: Sure I can do that for you!

**Comment by evrose54** [comment]
All of your conversations have been addressed!

File: geoips/create_plugin_registries.py
Comment (mindyls) [reply geoips/create_plugin_registries.py]: Looks good to me, as long as @jsolbrig is good with how this is laid out.  It is a lot of looping, but it is only done once at install time so is probably fine?

File: geoips/create_plugin_registries.py
Comment (mindyls) [reply geoips/create_plugin_registries.py]: As long as we make sure these error messages are raised appropriately when we integrate this with pyproject.toml, so if the installation fails it is obvious why, this is good with me.

File: geoips/create_plugin_registries.py
Comment (mindyls) [reply geoips/create_plugin_registries.py:146]: This looks good!  Lists both package names if there are duplicates found across 2 different plugin packages, and the single package name if there are duplicate plugins in the same plugin package!

**Review by mindyls — APPROVED** [review]
Looks good to me!

File: geoips/geoips_utils.py
Comment (jsolbrig) [inline geoips/geoips_utils.py:1]: Should this also test that the file exists and raise a PluginRegistryError if it doesn't at line 85?

File: geoips/create_plugin_registries.py
Comment (jsolbrig) [inline geoips/create_plugin_registries.py:84]: Can this list be derived from the registry files rather than hard-coded? We're likely going to add "text_based" in the near future and it is possible that we'll have others farther down the line, too. This would be one less thing to change if we can make it dynamic.

File: geoips/create_plugin_registries.py
Comment (jsolbrig) [reply geoips/create_plugin_registries.py:147]: This looks good for now. I expect that this will get moved into the PluginRegistry class in #389.

File: geoips/create_plugin_registries.py
Comment (jsolbrig) [reply geoips/create_plugin_registries.py]: Yeah, I think this looks good for now. I expect a lot of this functionality will move into `PluginRegistry` when we do #389.

File: geoips/create_plugin_registries.py
Comment (jsolbrig) [reply geoips/create_plugin_registries.py]: Looks good for now. We can handle the errors when we get to #385.

**Comment by jsolbrig** [comment]
This all looks good for now. I have a few thoughts, but I think they can be addressed when we do #389.

My thoughts:
- I think that a lot of this functionality should be moved into the PluginRegistry class when we do #389.
- I'm concerned that these functions are going to be tough to unit test. They take in a list of plugin repositories, then collect the registry files inside each function. I'd prefer to separate those duties. If functions like `registry_sanity_check` took in a list of registry files rather than a list of plugin packages, they would be much easier to test since we could pass in bad registries more easily.

File: tests/unit_tests/test_dtree.py
Comment (mindyls) [inline tests/unit_tests/test_dtree.py:1]: NOTE: I had to remove this file to get the pytests to run (there were duplicate files, here and in xarray_utils/test_dtree.py).  So the pytests will not consistently pass until this is merrge.

I am good with merging this as is, and addressing Jeremy's comments when his new issue is addressed.

File: geoips/create_plugin_registries.py
Comment (evrose54) [reply geoips/create_plugin_registries.py]: I've changed this to dynamically detect this.

Comment (jsolbrig) [reply geoips/create_plugin_registries.py]: Looks good. Thanks Evan.

**Pull request merged**
